### ISSUE TITLE
JMS Auto Instrumentation does not create a link to the publisher span on synschronous 'receive' when enabled

### ISSUE THREAD

**Issue by mgevantmakher** [open-telemetry/opentelemetry-java-instrumentation#6779]
Message 'receiver' span does not have any relationship to 'producer' created by JMS autinstrumentation. 
Consequence of not having any relationship between receiver span and the publisher trace is very limited message tracebility among service boundaries when JMS auto-instrumentation is used. 
Receiver trace is disabled by default, when enabled using 'otel.instrumentation.messaging.experimental.receive-telemetry.enabled=true' "receive' span is created and application span becomes a parent of it when available. There is no link available to the publisher span.  Kafka autoinstrumentation supports linking in such case. 

There is a PR to the messaging semantic conventions that describes a link in a batching use case, same concern is valid for single publisher: https://github.com/open-telemetry/oteps/blob/7c1d84ca415259f0cf2c6b7fd2dc7889dff4d9d6/text/trace/0192-messaging-semantic-conventions-spec.md#single-message-producers-batch-push-based-consumer-with-process-spans

To extract trace parent form a message the propagator api is usually used (for context propagation). It is not used right now since context propagation is not supported for 'receive'. If this feature will be implemented, the extension SPI should be able to replace a class that is responsible for trace parent information extraction (i.e propagator with a getter and setter). 

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/429ecfc7130419a49299968b3773097f040c7dfd/instrumentation/jms-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jms/JmsSingletons.java#L50

**Comment by mateuszrzeszutek** [comment]
> Receiver trace is disabled by default, when enabled using 'otel.instrumentation.messaging.experimental.receive-telemetry.enabled=true' "receive' span is created and application span becomes a parent of it when available. There is no link available to the publisher span. Kafka autoinstrumentation supports linking in such case.

The Kafka instrumentation does not add links to the consumer receive span either: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/5f75a5060bf5505901a930fde380fcfc706bfac0/instrumentation/kafka/kafka-clients/kafka-clients-common/library/src/main/java/io/opentelemetry/instrumentation/kafka/internal/KafkaInstrumenterFactory.java#L101-L114
It does add them to consumer process spans though.

The [current semantic conventions for batch receiving](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#batch-receiving) also don't have span links in the consumer receive span, only in the consumer process spans. This might change in the future though, as the messaging semantic conventions are currently being reworked and rewritten -- the OTEP that you posted is a part of that effort.

**Comment by mgevantmakher** [comment]
@mateuszrzeszutek Would you expect that this issue can be resolved in a near future Mateusz?

**Comment by mgevantmakher** [comment]
Luck of any kind of linking limits usability of the JMS autoinstrumentation for existing JMS applications substantially.

**Comment by mateuszrzeszutek** [comment]
Yeah, I took a look at the JMS instrumentation, it looks like the use case is slightly different from Kafka - it's either receive or process, not both at the same time (which is the case for Kafka). In that case, I agree that the receive span should link to the producer span -- without that the usability is rather poor. The semantic conventions don't describe a situation like that, but personally I think that adding that span link is probably the most sensible solution for now, because it will be a while before the conventions around messaging stabilize.
To sum up: let's add the link, it makes sense.

**Comment by mgevantmakher** [comment]
Thank you for supporting my view Mateusz, I will bring this up on a next SIG Java instrumentation meeting and see who can implement it and when. 
To add a link trace parent from a message needs to be extracted, so the default propagator/ overloadable using extension propagator needs to be involved.
